File: 17-03-2024/Count Pairs whose sum is equal to X.cs
//{ Driver Code Starts
//Initial Template for C#

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using DriverCode;

namespace DriverCode
{
    public class Node{
        public int data;
        public Node next;
        
        public Node(int a){
            this.data = a;
            this.next = null;
        }
        
    }
    
    
    class GFG
    {
        static void printList( Node head){
            while(head!=null){
                Console.Write(head.data+" ");
                head = head.next;
            }
            Console.Write("\n");
        }
        static void Main(string[] args)
        {
            int testcases;// Taking testcase as input
            testcases = Convert.ToInt32(Console.ReadLine());
            while (testcases-- > 0)// Looping through all testcases
            {
                string[] element = Console.ReadLine().Trim().Split(' ');
                int n = Convert.ToInt32(element[0]);
                string[] ndata = Console.ReadLine().Trim().Split(' ');
                int[] a = new int[n];
                a = Array.ConvertAll(ndata, int.Parse);
                
                Node head = new Node(a[0]);
                Node tail = head;
                
                for(int i=1;i<n;i++){
                    tail.next = new Node(a[i]);
                    tail = tail.next;
                }
                
                element = Console.ReadLine().Trim().Split(' ');
                n = Convert.ToInt32(element[0]);
                ndata = Console.ReadLine().Trim().Split(' ');
                a = new int[n];
                a = Array.ConvertAll(ndata, int.Parse);
                
                Node head1 = new Node(a[0]);
                tail = head1;
                
                for(int i=1;i<n;i++){
                    tail.next = new Node(a[i]);
                    tail = tail.next;
                }
                
                int p = Convert.ToInt32(Console.ReadLine());
                
                Solution obj = new Solution();
                Console.WriteLine(obj.countPairs(head, head1, p));
                
                
            }

        }
    }
    
    
}

// } Driver Code Ends




//User function Template for C#

// public class Node{
//     public int data;
//     public Node next;

//     public Node(int a){
//         this.data = a;
//         this.next = null;
//     }

// }
class Solution
    {
        //Complete this function
        public int countPairs(Node head1, Node head2, int x)
        {
          //Your Code Start
              HashSet<int> set = new HashSet<int>();

        // Traverse the second list and store its elements in a set
        Node temp = head2;
        while (temp != null)
        {
            set.Add(temp.data);
            temp = temp.next;
        }

        int count = 0;

        // Traverse the first list and for each element, check if its complement exists in the set
        temp = head1;
        while (temp != null)
        {
            if (set.Contains(x - temp.data))
            {
                count++;
            }
            temp = temp.next;
        }

        return count;
            //Your code End
        }
    }

File: 17-03-2024/Count Pairs whose sum is equal to X.js
//{ Driver Code Starts
//Initial Template for javascript

'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(string => {
        return string.trim();
    });
    
    main();    
});

function readLine() {
    return inputString[currentLine++];
}

class Node{
    constructor(data){
        this.data = data;
        this.next = null;
    }
}

function printlist(head){
    let current = head; 
    let s='';
    while (current !== null)
    {
        s+=current.data+" ";
        current = current.next;
    }
    console.log(s);
}


function main() {
    let t = parseInt(readLine());
    let i = 0;
    for(;i<t;i++)
    {
        let n = parseInt(readLine());
        let input_ar1 = readLine().split(' ').map(x=>parseInt(x));
        let head1 = new Node(input_ar1[0]);
        let tail1 = head1;
        for(let i=1;i<n;i++){
            tail1.next = new Node(input_ar1[i]);
            tail1 = tail1.next;
        }
        let m = parseInt(readLine());
        let input_ar2 = readLine().split(' ').map(x=>parseInt(x));
        let head2 = new Node(input_ar2[0]);
        let tail2 = head2;
        for(let i=1;i<m;i++){
            tail2.next = new Node(input_ar2[i]);
            tail2 = tail2.next;
        }
        let x = parseInt(readLine());
        let obj = new Solution();
        let res = obj.countPairs(head1, head2, x);
        console.log(res);
    }
}


// } Driver Code Ends




//User function Template for javascript

/**
 * @param {Node} head1
 * @param {Node} head2
 * @param {number} x
 * @returns {number}
*/

/*
class Node{
    constructor(data){
        this.data = data;
        this.next = null;
    }
}
*/

class Solution {
    
    countPairs(head1, head2, x)
    {
      //Your Code Start
         let set = new Set();

        // Traverse the second list and store its elements in a set
        let temp = head2;
        while (temp !== null) {
            set.add(temp.data);
            temp = temp.next;
        }

        let count = 0;

        // Traverse the first list and for each element, check if its complement exists in the set
        temp = head1;
        while (temp !== null) {
            if (set.has(x - temp.data)) {
                count++;
            }
            temp = temp.next;
        }

        return count;
        //your code End
    }
}
